drivers/stepper/gpio_stepper/gpio_stepper_controller.c

```diff
@@ -0,0 +1,205 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024 Carl Zeiss Meditec AG
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gpio_stepper
+
+#include "zephyr/drivers/gpio.h"
+#include <zephyr/kernel.h>
+#include <zephyr/sys_clock.h>
+#include "zephyr/drivers/stepper.h"
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/__assert.h>
+#include "stdlib.h"
+
+LOG_MODULE_REGISTER(gpio_stepper_motor_controller, CONFIG_STEPPER_LOG_LEVEL);
+
+#define NUMBER_OF_GPIO_PINS 4
+
+struct gpio_stepper_motor_controller_data {
+	struct k_spinlock lock;
+	struct k_work_delayable stepper_dwork;
+	const struct gpio_dt_spec *control_pins;
+	int32_t min_position;
+	int32_t max_position;
+	int32_t actual_position;
+	int32_t target_position;
+	uint32_t actual_speed;
+	uint32_t delay_in_us;
+};
+
+static int32_t stepper_motor_move_with_4_gpios(const struct gpio_dt_spec *gpio_spec,
+					       int step_number)
+{
+	switch (abs(step_number) % 4) {
```

Comment:
This calculation is quite computationally heavy (includes division). Wrapping the step_count conditionally at -1 and 4 before passing it to this function is faster.